File: client/src/main/java/com/nebula/client/utils/SignUtil.java
package com.nebula.client.utils;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.Map;
import java.util.TreeMap;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class SignUtil {

  private static final String HMAC_SHA256 = "HmacSHA256";
  private static final String SHA256 = "SHA-256";
  private static final int BUFFER_SIZE = 8192; // 8KB 缓冲区

  public static String genSign(String secretKey, String appId, String timestamp, String nonce, String body) {
    String sortedBody = sortBodyParameters(body); 
    String bodyHash = calculateBodyHash(sortedBody);
    String concatenatedString = buildConcatenatedString(appId, timestamp, nonce, bodyHash);
    return generateHmacSHA256(secretKey, concatenatedString);
  }

  private static String sortBodyParameters(String body) {
    if (body == null || body.isEmpty())
      return "";

    Map<String, String> params = new TreeMap<>();
    String[] pairs = body.split("&");
    for (String pair : pairs) {
      String[] keyValue = pair.split("=");
      if (keyValue.length == 2) {
        params.put(keyValue[0], keyValue[1]);
      }
    }
    StringBuilder sortedBody = new StringBuilder();
    for (Map.Entry<String, String> entry : params.entrySet()) {
      sortedBody.append(entry.getKey()).append("=").append(entry.getValue()).append("&");
    }
    return sortedBody.length() > 0 ? sortedBody.substring(0, sortedBody.length() - 1) : "";
  }

  private static String calculateBodyHash(String body) {
    if (body == null) {
      body = "";
    }
    try {
      MessageDigest digest = MessageDigest.getInstance(SHA256);
      try (InputStream inputStream = new ByteArrayInputStream(body.getBytes(StandardCharsets.UTF_8))) {
        byte[] buffer = new byte[BUFFER_SIZE];
        int bytesRead;
        while ((bytesRead = inputStream.read(buffer)) != -1) {
          digest.update(buffer, 0, bytesRead);
        }
      }
      byte[] hash = digest.digest();
      return bytesToHex(hash);
    } catch (NoSuchAlgorithmException | IOException e) {
      throw new RuntimeException("Failed to calculate SHA-256 hash", e);
    }
  }

  private static String buildConcatenatedString(String appId, String timestamp, String nonce, String bodyHash) {
    Map<String, String> params = new TreeMap<>();
    params.put("appId", appId);
    params.put("timestamp", timestamp);
    params.put("nonce", nonce);
    params.put("body", bodyHash);

    StringBuilder sb = new StringBuilder();
    for (Map.Entry<String, String> entry : params.entrySet()) {
      sb.append(entry.getKey())
          .append("=")
          .append(urlEncode(entry.getValue()))
          .append("&");
    }
    return sb.substring(0, sb.length() - 1); // 去掉最后一个 "&"
  }

  private static String urlEncode(String value) {
    try {
      return URLEncoder.encode(value, "UTF-8");
    } catch (UnsupportedEncodingException e) {
      throw new RuntimeException("Failed to encode URL", e);
    }
  }

  private static String generateHmacSHA256(String secretKey, String data) {
    try {
      Mac mac = Mac.getInstance(HMAC_SHA256);
      SecretKeySpec secretKeySpec = new SecretKeySpec(secretKey.getBytes(StandardCharsets.UTF_8), HMAC_SHA256);
      mac.init(secretKeySpec);
      byte[] signatureBytes = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
      return Base64.getEncoder().encodeToString(signatureBytes);
    } catch (Exception e) {
      throw new RuntimeException("Failed to generate HMAC-SHA256 signature", e);
    }
  }

  private static String bytesToHex(byte[] bytes) {
    StringBuilder sb = new StringBuilder();
    for (byte b : bytes) {
      sb.append(String.format("%02x", b));
    }
    return sb.toString();
  }
}

File: fronted/src/SignUtil.js
import crypto from "crypto"

export class SignUtil {
  static async genSign(secretKey, appId, timestamp, nonce, body) {
    const sortedBody = SignUtils.sortBody(body); // 对 body 进行排序
    const bodyHash = await SignUtils.calculateBodyHash(sortedBody); // 计算排序后的 body 的哈希值
    const concatenatedString = SignUtils.buildConcatenatedString(appId, timestamp, nonce, bodyHash);
    return SignUtils.generateHmacSHA256(secretKey, concatenatedString);
  }

  // 对 body 进行排序
  static sortBody(body) {
    if (!body) return '';
    const params = new URLSearchParams(body);
    params.sort(); // 对参数进行排序
    return params.toString();
  }

  // 计算 body 的 SHA-256 哈希值
  static async calculateBodyHash(body) {
    if (!body) body = '';
    const encoder = new TextEncoder();
    const data = encoder.encode(body);
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    return SignUtils.bytesToHex(new Uint8Array(hashBuffer));
  }

  // 构建签名字符串
  static buildConcatenatedString(appId, timestamp, nonce, bodyHash) {
    const params = new URLSearchParams();
    params.append('appId', appId);
    params.append('timestamp', timestamp);
    params.append('nonce', nonce);
    params.append('body', bodyHash);
    params.sort(); // 对参数进行排序
    return params.toString();
  }

  // 生成 HMAC-SHA256 签名
  static async generateHmacSHA256(secretKey, data) {
    const encoder = new TextEncoder();
    const key = await crypto.subtle.importKey(
      'raw',
      encoder.encode(secretKey),
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );
    const signature = await crypto.subtle.sign('HMAC', key, encoder.encode(data));
    return SignUtils.base64Encode(signature);
  }

  // 将字节数组转换为十六进制字符串
  static bytesToHex(bytes) {
    return Array.from(bytes)
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  // 将 ArrayBuffer 转换为 Base64 字符串
  static base64Encode(buffer) {
    const bytes = new Uint8Array(buffer);
    let binary = '';
    bytes.forEach((b) => (binary += String.fromCharCode(b)));
    return btoa(binary);
  }
}